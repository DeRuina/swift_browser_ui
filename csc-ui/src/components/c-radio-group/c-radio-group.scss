:host {
  display: flex;
  flex-direction: column;
  gap: 4px;
}

.c-radio-group {
  --c-radio-flex-direction: column;
  --c-radio-item-gap: 2px;

  display: flex;
  flex-direction: column;
  gap: 4px;
  width: fit-content;

  &__items {
    display: flex;
    flex-direction: var(--c-radio-flex-direction);
    flex-wrap: wrap;
    gap: var(--c-radio-item-gap);
  }

  &__label {
    span.required {
      color: var(--csc-error);
    }
  }

  &--disabled {
    --c-radio-color: var(--csc-mid-grey);
    --csc-error: var(--csc-mid-grey);

    color: var(--c-radio-color);
    cursor: default;
    opacity: 0.75;
  }

  &--error {
    --c-radio-color: var(--csc-error);

    color: var(--csc-error);
  }

  &--inline {
    --c-radio-flex-direction: row;
    --c-radio-item-gap: 12px;
  }
}

.c-radio {
  $this: &;

  --c-radio-color: var(--csc-primary);

  align-items: flex-start;
  display: flex;
  position: relative;
  cursor: pointer;
  font-size: 16px;
  user-select: none;
  gap: 4px;
  line-height: 1.2;

  input {
    position: absolute;
    opacity: 0;
    cursor: pointer;
    height: 0;
    width: 0;

    &:checked ~ .ripple .selection {
      &::after {
        display: block;
        transform: scale(1);
      }
    }
  }

  .ripple {
    position: relative;
    height: 42px;
    width: 42px;
    min-width: 42px;
    border-radius: 50%;
    overflow: hidden;
    transition: translateZ(0);
    color: var(--csc-primary);
  }

  .selection {
    position: absolute;
    top: 11px;
    left: 11px;
    height: 20px;
    width: 20px;
    background-color: transparent;
    box-shadow: inset 0 0 0 2px var(--c-radio-color);
    border-radius: 50%;
    transition: box-shadow 0.15s ease-in-out;

    &::after {
      content: '';
      position: absolute;
      transform: scale(0);
      transition: transform 0.15s ease-in-out;
      top: 5px;
      left: 5px;
      width: 10px;
      height: 10px;
      border-radius: 50%;
      background: var(--c-radio-color);
    }
  }

  &__details {
    line-height: 1;
    padding: 0 12px;

    &.active {
      #{$this} {
        &__message {
          opacity: 1;
          transform: translateY(0px);
        }
      }
    }
  }

  &__message {
    font-size: 12px;
    opacity: 0;
    transform: translateY(-4px);
    transition: all 0.2s cubic-bezier(0.25, 0.8, 0.5, 1);

    &--hint {
      color: var(--csc-mid-grey);
    }

    &--error {
      color: var(--csc-error) !important;
    }

    span {
      align-items: flex-start;
      display: flex;
      gap: 4px;
      height: 16px;

      &.error {
        color: var(--csc-error);
      }

      &.hint {
        color: var(--csc-mid-grey);
      }
    }

    svg {
      fill: currentColor;
      min-height: 16px;
      min-width: 16px;
      position: relative;
      top: -2px;
    }
  }

  &__label {
    padding-top: 12px;
  }

  &--disabled {
    --c-radio-color: var(--csc-mid-grey);

    color: var(--c-radio-color);
    cursor: default;
    opacity: 0.75;
  }

  &--error {
    --c-radio-color: var(--csc-error);

    color: var(--csc-error);

    span,
    svg {
      fill: var(--csc-error);
      color: var(--csc-error);
    }
  }

  input:focus {
    outline: none;
  }

  input:focus + .ripple {
    outline: 2px var(--c-radio-color) solid;
  }

  @supports selector(:focus-visible) {
    input:focus + .ripple {
      outline: none;
    }
  }

  input:focus-visible + .ripple {
    outline: 2px var(--c-radio-color) solid;
  }

  &:not(.c-radio--disabled) {
    .ripple:hover {
      background-color: var(--csc-primary-text-hover);
    }
  }
}
