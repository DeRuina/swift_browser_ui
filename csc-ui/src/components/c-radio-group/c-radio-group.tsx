import {
  Component,
  h,
  Prop,
  Event,
  EventEmitter,
  Element,
  State,
  Watch,
} from '@stencil/core';
import { mdiCloseCircle } from '@mdi/js';
import { CRadioGroupItem } from '../../types';
import { createRipple } from '../../utils/utils';

/**
 * @group Form
 * @slot - Default slot for the label
 */
@Component({
  tag: 'c-radio-group',
  styleUrl: 'c-radio-group.scss',
  shadow: true,
})
export class CRadioGroup {
  /**
   * Value of the radio group
   */
  @Prop({ mutable: true }) value: string | number | CRadioGroupItem;

  /**
   * Hide the hint and error messages
   */
  @Prop() hideDetails = false;

  /**
   * Hint text for the input
   */
  @Prop() hint = '';

  /**
   * Display radio buttons inline
   */
  @Prop() inline = false;

  /**
   * Label of the radio group
   */
  @Prop() label: string;

  /**
   * Color of the radio group
   */
  @Prop() color = '';

  /**
   * Radio group items
   */
  @Prop() items: CRadioGroupItem[] = [];

  /**
   * Disable the radio group
   */
  @Prop() disabled = false;

  /**
   * Return only the item value rather than the whole item object
   */
  @Prop() returnValue: false;

  /**
   * Set as required
   */
  @Prop() required = false;

  /**
   * Set the validity of the input
   */
  @Prop() valid = true;

  /**
   * Custom validation message
   */
  @Prop() validation = 'Required field';

  /**
   * Emit value change to the parent
   */
  @Event({ bubbles: false }) changeValue: EventEmitter;

  @Element() el: HTMLCRadioGroupElement;

  @State() messageOptions = {
    show: true,
    type: 'hint',
    content: '',
  };

  @Watch('validation')
  onValidationMessageChange(message: string) {
    this.onValidChange(message.length === 0);
  }

  @Watch('valid')
  onValidChange(valid: boolean) {
    this._handleValidation(valid || this.valid);
  }

  componentDidLoad() {
    this._handleValidation(this.valid, 0);
  }

  private _containers?: HTMLDivElement[] = [];

  private static _uniqueId = 0;

  private _handleKeyDown(event: KeyboardEvent, item, index) {
    if (['Space', 'Enter'].includes(event.code)) {
      event.preventDefault();

      this._select(event, item, index);
    }
  }

  private _handleValidation(valid: boolean, timeout = 200) {
    this.messageOptions = {
      ...this.messageOptions,
      show: false,
    };

    setTimeout(() => {
      this.messageOptions = {
        ...this.messageOptions,
        type: valid ? 'hint' : 'error',
        show: true,
        content: valid ? (
          <span>{this.hint}</span>
        ) : (
          <span>
            {this._validationIcon} {this.validation}
          </span>
        ),
      };
    }, timeout);
  }

  private _select(event, item, index) {
    if (this.disabled) return;

    createRipple(event, this._containers[index], true);
    this.value = this.returnValue ? item?.value : item;
    this.changeValue.emit(this.value);
  }

  private _getRadioButton = (item, index) => {
    const itemId = item.value.toString().replace(/[^a-zA-Z0-9-_]/g, '');
    const isChecked = this.returnValue
      ? this.items?.find((i) => i.value === item.value)?.value === this.value
      : this.value === item;

    const classes = {
      'c-radio': true,
      'c-radio--disabled': this.disabled,
      'c-radio--error': this.messageOptions.type === 'error',
    };

    return (
      <label
        class={classes}
        id={itemId}
        onKeyDown={(event) => this._handleKeyDown(event, item, index)}
      >
        <input
          type="radio"
          aria-checked={(this.value === item).toString()}
          aria-disabled={this.disabled.toString()}
          aria-labelledby={itemId}
          disabled={this.disabled}
          checked={isChecked}
          name={CRadioGroup._uniqueId.toString()}
          onChange={(event) => this._select(event, item, index)}
        />

        <span
          class="ripple"
          ref={(el) => (this._containers[index] = el as HTMLDivElement)}
        >
          <span class="selection"></span>
        </span>

        <div class="c-radio__label">{item.label}</div>
      </label>
    );
  };

  private _renderMessages() {
    if (this.hideDetails) return;

    const classes = {
      'c-radio__details': true,
      active: this.messageOptions.show,
    };

    const messageClasses = {
      'c-radio__message': true,
      [`c-radio__message--${this.messageOptions.type}`]: true,
    };

    return (
      <div class={classes}>
        <div class={messageClasses}>{this.messageOptions.content}</div>
      </div>
    );
  }

  private _validationIcon = (
    <svg height="16px" width="16px" viewBox="0 0 24 24">
      <path d={mdiCloseCircle} />
    </svg>
  );

  componentWillLoad() {
    CRadioGroup._uniqueId += 1;
  }

  render() {
    const slotHasContent = !!this.el.childNodes.length;

    const wrapperClasses = {
      'c-radio-group': true,
      'c-radio-group--disabled': this.disabled,
      'c-radio-group--inline': this.inline,
      'c-radio-group--error': this.messageOptions.type === 'error',
    };

    return (
      <div
        class={wrapperClasses}
        role="radiogroup"
        aria-labelledby="c-radio-group__label"
      >
        {(!!this.label || slotHasContent) && (
          <label class="c-radio-group__label">
            {!!this.label ? this.label : <slot></slot>}
            {this.required && <span class="required">&nbsp;*</span>}
          </label>
        )}

        <div class="c-radio-group__items">
          {this.items.map((item, index) => this._getRadioButton(item, index))}
        </div>

        {this._renderMessages()}
      </div>
    );
  }
}
