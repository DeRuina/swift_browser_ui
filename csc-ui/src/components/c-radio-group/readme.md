# c-radio



<!-- Auto Generated Below -->


## Properties

| Property      | Attribute      | Description                                                  | Type                                  | Default            |
| ------------- | -------------- | ------------------------------------------------------------ | ------------------------------------- | ------------------ |
| `color`       | `color`        | Color of the radio group                                     | `string`                              | `''`               |
| `disabled`    | `disabled`     | Disable the radio group                                      | `boolean`                             | `false`            |
| `hideDetails` | `hide-details` | Hide the hint and error messages                             | `boolean`                             | `false`            |
| `hint`        | `hint`         | Hint text for the input                                      | `string`                              | `''`               |
| `inline`      | `inline`       | Display radio buttons inline                                 | `boolean`                             | `false`            |
| `items`       | --             | Radio group items                                            | `CRadioGroupItem[]`                   | `[]`               |
| `label`       | `label`        | Label of the radio group                                     | `string`                              | `undefined`        |
| `required`    | `required`     | Set as required                                              | `boolean`                             | `false`            |
| `returnValue` | `return-value` | Return only the item value rather than the whole item object | `boolean`                             | `undefined`        |
| `valid`       | `valid`        | Set the validity of the input                                | `boolean`                             | `true`             |
| `validation`  | `validation`   | Custom validation message                                    | `string`                              | `'Required field'` |
| `value`       | `value`        | Value of the radio group                                     | `CRadioGroupItem \| number \| string` | `undefined`        |


## Events

| Event         | Description                     | Type               |
| ------------- | ------------------------------- | ------------------ |
| `changeValue` | Emit value change to the parent | `CustomEvent<any>` |


## Slots

| Slot | Description                |
| ---- | -------------------------- |
|      | Default slot for the label |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
