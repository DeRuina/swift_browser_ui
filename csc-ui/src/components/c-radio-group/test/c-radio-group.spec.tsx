import { newSpecPage } from '@stencil/core/testing';
import { CRadioGroup } from '../c-radio-group';
import { items } from './items';
import { h } from '@stencil/core';

describe('c-radio-group', () => {
  it('renders default html', async () => {
    const page = await newSpecPage({
      components: [CRadioGroup],
      template: () => <c-radio-group items={items}></c-radio-group>,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('renders custom color and items inlined', async () => {
    const page = await newSpecPage({
      components: [CRadioGroup],
      template: () => (
        <c-radio-group color="red" inline items={items}></c-radio-group>
      ),
    });

    expect(page.root).toMatchSnapshot();
  });

  it('renders hint and label', async () => {
    const page = await newSpecPage({
      components: [CRadioGroup],
      template: () => (
        <c-radio-group
          hint="This is hint message"
          label="This is label"
          items={items}
        ></c-radio-group>
      ),
    });

    expect(page.root).toMatchSnapshot();
  });
});
