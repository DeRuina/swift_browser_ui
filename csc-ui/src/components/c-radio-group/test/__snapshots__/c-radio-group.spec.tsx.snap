// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`c-radio-group renders custom color and items inlined 1`] = `
<c-radio-group>
  <mock:shadow-root>
    <div aria-labelledby="c-radio-group__label" class="c-radio-group c-radio-group--inline" role="radiogroup">
      <div class="c-radio-group__items">
        <label class="c-radio" id="one">
          <input aria-checked="false" aria-disabled="false" aria-labelledby="one" name="2" type="radio">
          <span class="ripple">
            <span class="selection"></span>
          </span>
          <div class="c-radio__label">
            Radio 1
          </div>
        </label>
        <label class="c-radio" id="two">
          <input aria-checked="false" aria-disabled="false" aria-labelledby="two" name="2" type="radio">
          <span class="ripple">
            <span class="selection"></span>
          </span>
          <div class="c-radio__label">
            Radio 2
          </div>
        </label>
        <label class="c-radio" id="three">
          <input aria-checked="false" aria-disabled="false" aria-labelledby="three" name="2" type="radio">
          <span class="ripple">
            <span class="selection"></span>
          </span>
          <div class="c-radio__label">
            Radio 3
          </div>
        </label>
      </div>
      <div class="c-radio__details">
        <div class="c-radio__message c-radio__message--hint"></div>
      </div>
    </div>
  </mock:shadow-root>
</c-radio-group>
`;

exports[`c-radio-group renders default html 1`] = `
<c-radio-group>
  <mock:shadow-root>
    <div aria-labelledby="c-radio-group__label" class="c-radio-group" role="radiogroup">
      <div class="c-radio-group__items">
        <label class="c-radio" id="one">
          <input aria-checked="false" aria-disabled="false" aria-labelledby="one" name="1" type="radio">
          <span class="ripple">
            <span class="selection"></span>
          </span>
          <div class="c-radio__label">
            Radio 1
          </div>
        </label>
        <label class="c-radio" id="two">
          <input aria-checked="false" aria-disabled="false" aria-labelledby="two" name="1" type="radio">
          <span class="ripple">
            <span class="selection"></span>
          </span>
          <div class="c-radio__label">
            Radio 2
          </div>
        </label>
        <label class="c-radio" id="three">
          <input aria-checked="false" aria-disabled="false" aria-labelledby="three" name="1" type="radio">
          <span class="ripple">
            <span class="selection"></span>
          </span>
          <div class="c-radio__label">
            Radio 3
          </div>
        </label>
      </div>
      <div class="c-radio__details">
        <div class="c-radio__message c-radio__message--hint"></div>
      </div>
    </div>
  </mock:shadow-root>
</c-radio-group>
`;

exports[`c-radio-group renders hint and label 1`] = `
<c-radio-group>
  <mock:shadow-root>
    <div aria-labelledby="c-radio-group__label" class="c-radio-group" role="radiogroup">
      <label class="c-radio-group__label">
        This is label
      </label>
      <div class="c-radio-group__items">
        <label class="c-radio" id="one">
          <input aria-checked="false" aria-disabled="false" aria-labelledby="one" name="3" type="radio">
          <span class="ripple">
            <span class="selection"></span>
          </span>
          <div class="c-radio__label">
            Radio 1
          </div>
        </label>
        <label class="c-radio" id="two">
          <input aria-checked="false" aria-disabled="false" aria-labelledby="two" name="3" type="radio">
          <span class="ripple">
            <span class="selection"></span>
          </span>
          <div class="c-radio__label">
            Radio 2
          </div>
        </label>
        <label class="c-radio" id="three">
          <input aria-checked="false" aria-disabled="false" aria-labelledby="three" name="3" type="radio">
          <span class="ripple">
            <span class="selection"></span>
          </span>
          <div class="c-radio__label">
            Radio 3
          </div>
        </label>
      </div>
      <div class="c-radio__details">
        <div class="c-radio__message c-radio__message--hint"></div>
      </div>
    </div>
  </mock:shadow-root>
</c-radio-group>
`;
